Add individual documentation pages for string's subcommands

krobelus opened the pull request:
This adds string-x.rst for each subcommand x of string.  Apart from a
short description in the header, everything in string-x.rst is
generated based on the subcommand section in string.rst.

It can be viewed with either of:

```
man string collect
man string-collect
string collect<Alt-h>
string collect -h
```

While `string -h collect` still prints the full help.

Closes #5968


----

## TODOs:
<!-- Just check off what what we know been done so far. We can help you with this stuff. -->
- [ ] User-visible changes noted in CHANGELOG.md
- [x] Add the examples from string.rst as well

faho commented on sphinx_doc_src/fish_sphinxext.py:
```diff
@@ -0,0 +1,82 @@
+
+def string_doc_lines(start, stop):
+    """Print lines from the string documentation start from the line that matches start
+    and stopping tat the first line that matches stop."""
```
Typo: "tat" should be "at"

faho commented on sphinx_doc_src/fish_sphinxext.py:
```diff
@@ -0,0 +1,82 @@
+from docutils.parsers import rst
+from docutils.statemachine import ViewList
+
+STRING_DOC = os.path.join(os.path.dirname(__file__), "cmds/string.rst")
```
I think the naming overdoes it a bit with `string` everywhere. It's entirely plausible we'll split more docs in future.

faho commented on sphinx_doc_src/fish_sphinxext.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def string_doc_lines(start, stop):
+    """Print lines from the string documentation start from the line that matches start
```
Like this - there is nothing here specific to `string`. It's a function that matches between two regexes (which is also a string command I've been toying with)

faho commented on sphinx_doc_src/fish_sphinxext.py:
```diff
@@ -0,0 +1,82 @@
+        if not result:
+            if re.match(start, line):
+                i += 1
+                result += [line, STRING_DOC_LINES[i]]  # always include the header's ---
```
Shouldn't you be able to get by without the `i += 1` here if you just add the first line?

That would also then allow you to rewrite it as a `for` loop which is more idiomatic here.

Something like

```python
    result = []
    for line in STRING_DOC_LINES:
        if not result:
            if re.match(start, line):
                result += [line]
        elif re.match(stop, line):
                return result
        else:
            result += [line]
```

krobelus replied:
This is a hack - for the subcommand description I want to start at a header line and stop at a line with all hyphens - but not the one immediately after the header. We should not need this logic if we restructure the doc sources.

zanchey commented:
This fails to build for me:

```
/Users/david/src/fish-shell/src/builtin_string.cpp:1377:9: error: no matching function for call to 'builtin_print_help'
        builtin_print_help(parser, streams, string_dash_subcommand.c_str(), streams.out);
        ^~~~~~~~~~~~~~~~~~
/Users/david/src/fish-shell/src/builtin.h:91:6: note: candidate function not viable: no known conversion from 'output_stream_t'
      to 'wcstring *' (aka 'basic_string<wchar_t, char_traits<wchar_t>, allocator<wchar_t> > *') for 4th argument
void builtin_print_help(parser_t &parser, io_streams_t &streams, const wchar_t *cmd,
     ^
```

I am not sure that the major increase in complexity here is worth it.

krobelus commented:
Rebased and fixed build.
(Currently this provokes some warnings from sphinx, saying `Unexpected section title.`)

> I am not sure that the major increase in complexity here is worth it.

The man pages for subcommands are much less overwhelming; not sure if it matters in the HTML version. I understand this script is brittle, and the generated subcommand docs do not even include example sections. Maybe a different approach can reduce the complexity: move the documentation for `string foo` from `string.rst` to `string-foo.rst` (keeping only generic docs in `string.rst`).
Then we can decide on whether to
1. include the full subcommand docs in the page for `string` (which should be simpler than the current approach).
2. merely link to the subcommands (then we don't need any sphinx extension).

Doing this would enable us to show the examples for some subcommand right after the description. It may be less convenient to maintain many small files, though. Plus we might have to move some things around to have a separate example section for each subcommand.

zanchey commented:
That sounds better to me.

krobelus commented:
Pushed the new version. Now string.rst includes the string-collect.rst and so on using Sphinx' `include`, no python extension necessary.

The only thing that this changes in the main string page is that the examples for each subcommand are included immediately after the subcommand description. This can be changed easily.

 The use is mainly to have man pages for string subcommands; we could disable the HTML outputs for the subcommand pages, since all the info is in string.html.

faho commented:
>The only thing that this changes in the main string page is that the examples for each subcommand are included immediately after the subcommand description. This can be changed easily.

I don't think it needs to be.

In fact this LGTM as it is.
